### PR TITLE
Add retries for get_workflow_job_id and try catch in upload_test_stats

### PR DESCRIPTION
**Pull request by clee2000**
upload_test_stats keeps failing b/c it can't handle when the id is workflow-<workflow_id> so add a try catch for this.

Add retries to get_workflow_job_id to try and reduce the number of times the id can't be found

Failure to upload test stats and inability to get the job id cause our sharding infra and slow test infra (probably also flaky test detection) to be less effective.  This does not completely resolve the issue since we do rely on the job id

Failure to get the workflow job id happens tragically often, hopefully retries will help

### CODE REVIEW

**Review by huydhn — APPROVED** [review]
The change itself looks good to me.  But I wonder if it GitHub infra is that bad that retrying here makes any difference (I assume network flakiness happens but rarely)

Also adding retry in the workflow https://github.com/pytorch/pytorch/blob/master/.github/actions/get-workflow-job-id/action.yml#L18 looks easier if the script returns non-zero status code correctly

File: .github/scripts/get_workflow_job_id.py [hunk 1]
```diff
@@ -37,16 +37,19 @@ def fetch_url(url: str, *,
...
+    for i in range(retries + 1):
+        try:
+            with urlopen(Request(url, headers=headers)) as conn:
+                return reader(conn)
+        except urllib.error.HTTPError as err:
+            exception_message = (
+                "Is github alright?",
+                f"Recieved status code '{err.code}' when attempting to retrieve {url}:\n",
+                f"{err.reason}\n\nheaders={err.headers}"
+            )
```
Comment (seemethere) [inline .github/scripts/get_workflow_job_id.py:50]: Should we add a sleep here?

File: .github/scripts/get_workflow_job_id.py [hunk 2]
```diff
@@ -37,16 +38,20 @@ def fetch_url(url: str, *,
...
-    except urllib.error.HTTPError as err:
-        exception_message = (
-            "Is github alright?",
-            f"Recieved status code '{err.code}' when attempting to retrieve {url}:\n",
-            f"{err.reason}\n\nheaders={err.headers}"
-        )
-        raise RuntimeError(exception_message) from err
+    retries = 3
+    for i in range(retries + 1):
+        try:
```
Comment (malfet) [inline .github/scripts/get_workflow_job_id.py:43]: Sorry for the late review, but this change does two things which are imo bad from design point of view:
 - Hardcodes constants
 - Builds [Pyramid of Doom](https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming))
Let me propose a PR that fixes the problem

Comment (malfet) [reply .github/scripts/get_workflow_job_id.py:43]: https://github.com/pytorch/pytorch/pull/93874

**Pull request closed without merging**